### PR TITLE
implement decay

### PR DESCRIPTION
**Pull request by soenkehahn**
cc: @dan-f

I've implemented decay. I'd appreciate a review, since I think there's probably some parts that could be done better. One note: When an oscillator is in the state of `Decaying` and you trigger a new note, there will still be an audible click, since it'll reset the phase. The proper solution would be to implement an `Attack` state (as in ADSR), but that's a bit more involved. So I would defer that.

### CODE REVIEW

File: src/run_jack.rs [hunk 1]
```diff
@@ -5,13 +5,16 @@ use jack::*;
 use std::sync::{Arc, Mutex};
 use std::*;
 
-pub fn run_jack_generator(
-    name: String,
-    generator: Arc<Mutex<Generator>>,
-) -> Result<AsyncClient<(), ProcessHandler_>, Error> {
+pub fn make_client(name: String) -> Result<Client, Error> {
```
Comment (dan-f) [inline src/run_jack.rs]: "Simplify to one function returning a struct that owns the `AsyncClient` and that allows you to set frequency, etc." - said the wise @soenkehahn

File: src/generator.rs [hunk 2]
```diff
@@ -25,43 +35,109 @@ impl Generator {
             frequency,
             phase: match self.oscillator_state {
                 OscillatorState::Playing { phase, .. } => phase,
+                OscillatorState::Decaying { .. } => 0.0,
```
Comment (dan-f) [inline src/generator.rs:36]: Note: in current monophonic system, resets to phase during decay will trade off audible clicks during decay for having the proper starting phase of the new note.

File: src/generator.rs [hunk 3]
```diff
@@ -25,43 +35,109 @@ impl Generator {
...
                 OscillatorState::Muted => 0.0,
             },
         };
     }
 
     pub fn mute(&mut self) {
```
Comment (dan-f) [inline src/generator.rs]: I wonder if the name of this should change to something like "note_off" to indicate that it's not an immediate mute, but respects note decay.

File: src/generator.rs [hunk 4]
```diff
@@ -25,43 +35,109 @@ impl Generator {
...
         match self.oscillator_state {
             OscillatorState::Playing {
                 frequency,
                 ref mut phase,
             } => {
-                *phase += frequency * TAU / sample_rate as f32;
-                *phase %= TAU;
+                crank(sample_rate, frequency, phase);
+            }
+            OscillatorState::Decaying {
```
Comment (dan-f) [inline src/generator.rs]: @soenkehahn note to self - perhaps we can "or" these two patterns together?

File: src/generator.rs [hunk 5]
```diff
@@ -25,43 +35,109 @@ impl Generator {
...
+        if let OscillatorState::Decaying {
+            ref mut decay_amplitude,
+            ..
+        } = self.oscillator_state
+        {
+            *decay_amplitude -= self.decay_per_sample;
+        }
+        if let OscillatorState::Decaying {
+            decay_amplitude, ..
+        } = self.oscillator_state
```
Comment (dan-f) [inline src/generator.rs]: Note - repeatedly borrowing is a bit awkward, but possibly necessary?

Comment (soenkehahn) [reply src/generator.rs]: I've come up with a solution that is better than what we had before. It's still a bit cumbersome, but already much better.

File: src/generator.rs (hunk 4 above)
Comment (soenkehahn) [reply src/generator.rs]: Yup, works: https://github.com/soenkehahn/rust-device-reading/pull/3/commits/01a0602999a23e39203ee8152ed32b5e21341782

:)

File: src/generator.rs (hunk 3 above)
Comment (soenkehahn) [reply src/generator.rs]: Done.

File: src/run_jack.rs (hunk 1 above)
Comment (soenkehahn) [reply src/run_jack.rs]: Done.

**Comment by soenkehahn** [comment]
@dan-f: Thanks a lot for the review. I'm merging now.

**Pull request merged**